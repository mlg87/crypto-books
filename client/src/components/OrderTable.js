import React, { Component } from 'react'
import PropTypes from 'prop-types'
// mui
import CircularProgress from 'material-ui/Progress/CircularProgress'
import Typography from 'material-ui/Typography'
import Table, {
  TableBody,
  TableCell,
  TableHead,
  TableRow
} from 'material-ui/Table'
import green from 'material-ui/colors/green'
// lodash
import upperFirst from 'lodash/upperFirst'

export default class OrderTable extends Component {
  static propTypes = {
    currentMarket: PropTypes.string.isRequired,
    data: PropTypes.array.isRequired,
    denseCells: PropTypes.bool,
    showTitle: PropTypes.bool,
    isUpdatingTable: PropTypes.bool,
    pointOfComparison: PropTypes.number.isRequired,
    type: PropTypes.string.isRequired
  }

  static defaultProps = {
    denseCells: false,
    showTitle: false,
    isUpdatingTable: true
  }

  isOverlapped(price) {
    const { pointOfComparison, type } = this.props
    const actions = {
      asks: () => price < pointOfComparison,
      bids: () => price > pointOfComparison
    }
    return actions[type] ? actions[type]() : false
  }

  render() {
    const {
      currentMarket,
      data,
      denseCells,
      showTitle,
      isUpdatingTable,
      type
    } = this.props

    return (
      <div>
        {showTitle && (
          <Typography align="center" gutterBottom type="display1">
            {upperFirst(type)}
          </Typography>
        )}
        {isUpdatingTable ? (
          <div style={styles.loading.container}>
            <CircularProgress style={styles.loading.indicator} />
          </div>
        ) : (
          <div style={styles.table.container}>
            <Table>
              <TableHead>
                <TableRow>
                  <TableCell padding={denseCells ? 'dense' : 'default'}>
                    Exchange
                  </TableCell>
                  <TableCell
                    numeric
                    padding={denseCells ? 'dense' : 'default'}
                  >{`Price (${currentMarket})`}</TableCell>
                  <TableCell numeric padding={denseCells ? 'dense' : 'default'}>
                    Amount
                  </TableCell>
                </TableRow>
              </TableHead>
              <TableBody>
                {data.map((order, i) => {
                  const { amount, exchange, price } = order
                  return (
                    <TableRow
                      key={`title${i}`}
                      style={
                        this.isOverlapped(price) ? styles.table.overlap : null
                      }
                    >
                      <TableCell padding={denseCells ? 'dense' : 'default'}>
                        {upperFirst(exchange)}
                      </TableCell>
                      <TableCell
                        numeric
                        padding={denseCells ? 'dense' : 'default'}
                      >
                        {price}
                      </TableCell>
                      <TableCell
                        numeric
                        padding={denseCells ? 'dense' : 'default'}
                      >
                        {amount}
                      </TableCell>
                    </TableRow>
                  )
                })}
              </TableBody>
            </Table>
          </div>
        )}
      </div>
    )
  }
}

const styles = {
  table: {
    container: {
      maxHeight: '800px',
      overflow: 'scroll'
    },
    overlap: {
      backgroundColor: green[500]
    }
  },
  loading: {
    container: {
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      alignItems: 'center',
      marginTop: '40px'
    },
    indicator: {
      color: '#fff'
    }
  }
}
