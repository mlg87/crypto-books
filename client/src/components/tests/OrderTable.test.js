import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import OrderTable from '../OrderTable'

const props = {
  currentMarket: 'ETH/BTC',
  data: [1, 2, 3],
  pointOfComparison: 1,
  type: 'bid'
}

describe('OrderTable', () => {
  test('renders correctly', () => {
    const tree = renderer.create(<OrderTable {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
