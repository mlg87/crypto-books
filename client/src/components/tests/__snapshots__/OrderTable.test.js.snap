// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderTable renders correctly 1`] = `
<div>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "marginTop": "40px",
      }
    }
  >
    <div
      className="MuiCircularProgress-root-1 MuiCircularProgress-colorPrimary-2"
      role="progressbar"
      style={
        Object {
          "color": "#fff",
          "height": 40,
          "width": 40,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svgIndeterminate-4"
        viewBox="0 0 50 50"
      >
        <circle
          className="MuiCircularProgress-circle-6 MuiCircularProgress-circleIndeterminate-7"
          cx={25}
          cy={25}
          fill="none"
          r={20}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </div>
  </div>
</div>
`;
